### ISSUE TITLE
Create/ update raster pass observation type field as url instead of code

### ISSUE THREAD

**Issue by Derryrover** [nens/lizard-management-client#468]
Create/ update raster pass observation type field as url instead of code
See comments here:
https://github.com/nens/lizard-nxt/pull/4527

Currently we pass the code of the observation type, but in the new version of the backend this should be the url.

**Comment by caspervdw** [comment]
It should be one of three things, but not the UUID:

- https://nxt3.staging.lizard.net/api/v4/observationtypes/476/
- /api/v4/observationtypes/476/
- 476

If you put the URL (first or second option), that is currently already supported. So then the management client release can go before the backend.

**Comment by caspervdw** [comment]
Regarding the planning: this will for sure not go into the production release of 15 October. I estimate it will be November at the earliest.

Staging release can be any moment now.

**Comment by Derryrover** [comment]
Problem:

currently on staging If I send this:
{"name":"Tom test observation type api new 1","organisation":"61f5a464c35044c19bc7d4b42d7f58cb","observation_type":"https://nxt3.staging.lizard.net/api/v3/observationtypes/481/","description":"Tom test observation type api new 1","supplier":null,"aggregation_type":"none","options":{"styles":"3di-damage"},"shared_with":[],"rescalable":false,"access_modifier":"Private","temporal":false,"source":{"graph":{"raster":["lizard_nxt.blocks.LizardRasterSource","3b76efb6-062f-4a8c-a688-2d6c58fb3517"]},"name":"raster"}}

I receive this:
{"status":400,"code":20,"message":"Incomplete request content. #400.20","detail":{"observation_type":["Nested object 'https://nxt3.staging.lizard.net/api/v3/observationtypes/481/' does not exist. We do not allow the creation of nested objects."]}}

The old situation of simply sending the code of the observation type in this case "WNDDIR" did still work.

**Comment by Derryrover** [comment]
@caspervdw 

Hi Casper,

Now we have to coordinate backend + frontend releases anyway I wonder the following:

- Is there a reason that we not use the uuid when posting a relation to a foreign table (in this case raster to observation-type)? This uuid field is always unique and we might add this field to all our tables. This way all (or most) foreign relations in the api can be consistently a uuid. Takes a burden from the users of the api, they no longer have to look it up case by case. And takes a burden from the designers of the api, because you no longer have to decide it for each case seperately.

- If we stick to the existing situation and I have to choose between those:
    https://nxt3.staging.lizard.net/api/v4/observationtypes/476/
    /api/v4/observationtypes/476/
    476
   Which one would you prefer me to use?

We are doing a major overhaul on the management client right now.

kr Tom

**Comment by hoanphungt** [comment]
@Derryrover 
Hi Tom,

I make this PR: https://github.com/nens/lizard-management-client/pull/491 to use observation type's id when creating or updating a raster. This is now working on staging but not on production. 

Could you review the PR? Then we can deploy it to staging to test and later on, coordinate with the backend for the production release.

@caspervdw When do you expect to have the new implementation on production?

**Comment by Derryrover** [comment]
@hoanphungt 

Hi Hoan,
I tested and reviewed the code.
Ready to merge
kr Tom

**Comment by hoanphungt** [comment]
Released to staging version 0.2.5

**Comment by caspervdw** [comment]
According to @WolfSchouten  it does not work yet on staging. @hoanphungt  can you check maybe?

**Comment by WolfSchouten** [comment]
The management client sends an API POST with `"observation_type":"index"`, which gives this response `{"status":400,"code":20,"message":"Incomplete request content. #400.20","detail":{"observation_type":["Invalid hyperlink - Object does not exist."]}}`

When modifying the API POST to send `"observation_type": "4883"` it is succesful.

**Comment by caspervdw** [comment]
I think I'll add a (temporary) backwards compatibility here that it keeps accepting slugs. SO that we don't need to coordinate releases.

**Comment by hoanphungt** [comment]
@WolfSchouten @caspervdw 
Strange, I can create new raster or update an existing raster on staging without any error. Just checked it again and it still worked. I believe @Derryrover also tested it yesterday and it also worked for him.

**Comment by hoanphungt** [comment]
@WolfSchouten 
Can you explain how you tested this? maybe we did different things

**Comment by WolfSchouten** [comment]
Go here: https://nxt3.staging.lizard.net/management/#/data_management/rasters/new

Fill in all the required fields and click submit. While monitoring the connections in the dev tools network tab.

![image](https://user-images.githubusercontent.com/48123376/98231089-429d7600-1f5c-11eb-9e3f-676c175a1465.png)
Is the screen that the client shows.

**Comment by hoanphungt** [comment]
@WolfSchouten 
I did the same. Go there, fill in all the fields and click submit. The response is different for me though as I got this:
<img width="922" alt="Screenshot 2020-11-05 at 11 48 01" src="https://user-images.githubusercontent.com/43402556/98231649-fb63b500-1f5c-11eb-9305-ffc248505d8e.png">

**Comment by hoanphungt** [comment]
@WolfSchouten 
Can you try to hard reset the page to see if it is gonna work? Probably error because of cached files I think

**Comment by WolfSchouten** [comment]
Did a test in a private browsing window to make sure no cached files would be used. Now it does work as expected.

And I can see in the Request Payload that indeed the  key is sent, instead of the slug.

Closing this issue. And thank you for your insight.

**Comment by WolfSchouten** [comment]
Uh, maybe it shouldn't have been closed?

**Comment by caspervdw** [comment]
Maybe the browser caching of old clients is an issue in itself?

**Comment by hoanphungt** [comment]
@WolfSchouten no, it should only be closed after deployed to production succesfully :). 

@caspervdw I think browser caching of old management client is indeed an issue as I experienced it quite often but probably it has more to do with server setting than the front-end.

**Comment by Derryrover** [comment]
@hoanphungt 
@caspervdw 

Here is how we did it for 3di livesite:
https://github.com/nens/threedi-live-site/issues/570
Unfortunately this ticket is copied from JIRA so we miss a lot of detials.

Anyway what I remember:
- we should find out if it is the index.html that was loaded from cache or one of the linked javascript files. The linked javascript files should be not loaded from cache right now, because they get a new name every new build. Therefore I expect that it is the index.html that is loaded from cache. (If it is in contrast the javascript file that is cached, then we need to change something in our build process for frontend)
- If it is indeed the index.html then a solution is here:
https://stackoverflow.com/questions/1341089/using-meta-tags-to-turn-off-caching-in-all-browsers
In short the caching should not be done by the html file itself but by one of the proxies like NGINX. This is also what we did for the livesite eventually.
How exactly that is done I do not know, but Jelle will know it.

**Comment by Derryrover** [comment]
@joeriverheijden 
@caspervdw 

Management-client on staging and prod is same version:
Release 0.2.3

I just tested on staging to create a new raster and update the observation type. Both works.
I confirm there is for management-client no action required for the release of 9 december 2020.

**Comment by Derryrover** [comment]
additional note: Release 0.2.5 will also work correct after release of 9 devember 2020, but no need to update to that one.
